applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.cpp

```diff
@@ -0,0 +1,603 @@
+        }
+    }
+        else
+    {
+        rOutput.clear();
+    }
+
+    KRATOS_CATCH("")
+}
+
+void AdjointFiniteDifferencingBaseElement::GetValueOnIntegrationPoints(const Variable<double>& rVariable,
+                    std::vector<double>& rValues,
+                    const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    if(this->Has(rVariable))
+    {
+        // Get result value for output
+        double output_value = this->GetValue(rVariable);
+
+        // Resize Output
+        const unsigned int&  write_points_number = GetGeometry()
+            .IntegrationPointsNumber(this->GetIntegrationMethod());
+        if (rValues.size() != write_points_number)
+        {
+            rValues.resize(write_points_number);
+        }
+
+        // Write scalar result value on all Gauss-Points
+        for(unsigned int i = 0; i < write_points_number; ++i)
+        {
+            rValues[i] = output_value;
+        }
+    }
+    else
+        KRATOS_ERROR << "Unsupported output variable." << std::endl;
+
+    KRATOS_CATCH("")
+}
+
+int AdjointFiniteDifferencingBaseElement::Check(const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    GeometryType& r_geom = GetGeometry();
+
+    // verify that the variables are correctly initialized
+    KRATOS_CHECK_VARIABLE_KEY(DISPLACEMENT);
+    KRATOS_CHECK_VARIABLE_KEY(ROTATION);
+    KRATOS_CHECK_VARIABLE_KEY(VELOCITY);
+    KRATOS_CHECK_VARIABLE_KEY(ACCELERATION);
+    KRATOS_CHECK_VARIABLE_KEY(DENSITY);
+    KRATOS_CHECK_VARIABLE_KEY(CONSTITUTIVE_LAW);
+    KRATOS_CHECK_VARIABLE_KEY(ADJOINT_DISPLACEMENT);
+    KRATOS_CHECK_VARIABLE_KEY(ADJOINT_ROTATION);
+
+    // TODO generic way of doing these checks without checking the dofs..
+
+    // Check dofs
+    for (unsigned int i = 0; i < r_geom.size(); i++)
+    {
+        auto& r_node = r_geom[i];
+
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(DISPLACEMENT, r_node);
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(ROTATION, r_node);
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(ADJOINT_DISPLACEMENT, r_node);
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(ADJOINT_ROTATION, r_node);
+
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_X, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_Y, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_Z, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_ROTATION_X, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_ROTATION_Y, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_ROTATION_Z, r_node);
+    }
+
+    return 0;
+
+    KRATOS_CATCH("")
+}
+
+// Sensitivity functions
+
+void AdjointFiniteDifferencingBaseElement::CalculateSensitivityMatrix(const Variable<double>& rDesignVariable, Matrix& rOutput,
+                                            const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    // define working variables
+    Vector RHS_undist;
```

Comment:
If one does not know that this means undisturbed then it is tricky to figure it out
=> why not `RHS_undisturbed`?

Reply:
Done - disturbed is changed to perturbed